### PR TITLE
Persist bundle bytes to disk

### PR DESCRIPTION
**Pull request by ashutosh-narkar**
Earlier with bundle persistence enabled, the bundle
plugin would save the bundle object to disk. In
scenarios where the downloaded bundle has multiple
data files, OPA would first read the bundle and merge
data in the bundle under the bundle.Bundle struct's
Data field. Then before persisting the bundle to disk,
the bundle plugin would use the bundle writer to write
the bundle to the provided output stream. The result
of this is that all the data files in the original
bundle are consolidated into one data.json file.

Now if signature verification is enabled, it will fail
since the files includes in the bundle's signature will not
match the ones in the persisted bundle.

This commit resolves this issue by persiting the bytes
of downloaded bundle to disk which then loaded
from disk maintain the same structure as the original.

Fixes: #3472

Signed-off-by: Ashutosh Narkar <anarkar4387@gmail.com>

### CODE REVIEW

File: download/download.go [hunk 1]
```diff
@@ -183,17 +193,17 @@ func (d *Downloader) loop(ctx context.Context) {
 
 func (d *Downloader) oneShot(ctx context.Context) (bool, error) {
 	m := metrics.New()
-	b, etag, longPoll, err := d.download(ctx, m)
+	b, etag, longPoll, r, err := d.download(ctx, m)
```
Comment (srenatus) [inline download/download.go:196]: 💭 Time for a struct, maybe?

File: plugins/bundle/plugin.go [hunk 2]
```diff
@@ -548,21 +549,24 @@ func (p *Plugin) saveBundleToDisk(name string, b *bundle.Bundle) error {
...
+	if raw == nil {
+		return fmt.Errorf("no raw bundle bytes to persist to disk")
+	}
+
+	b, err := ioutil.ReadAll(raw)
+	if err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(filepath.Join(path, filename), b, 0644)
```
Comment (srenatus) [inline plugins/bundle/plugin.go:569]: [nit] this looks like we're bending ourselves to use ioutil.WriteFile. If instead we'd use `os.Open` on the file, we could use `io.Copy(dst io.Writer, src io.Reader)` to shovel the data into the file. What do you think?

File: plugins/bundle/plugin_test.go [hunk 3]
```diff
@@ -2082,6 +2203,19 @@ func getTestSignedBundle(t *testing.T) bundle.Bundle {
...
+
+	b := getTestBundle(t)
+
+	var buf bytes.Buffer
+	if err := bundle.NewWriter(&buf).UseModulePath(true).Write(b); err != nil {
+		t.Fatal("unexpected error:", err)
+	}
+
+	return &buf
+}
```
Comment (srenatus) [inline plugins/bundle/plugin_test.go:2217]: Looks like we could use this helper a bit more often above? Or am I missing something?

**Review by srenatus — DISMISSED** [review]
Nitpicks only.

I've been pondering alternatives to the current return-the-reader approach which seemed a bit indirect, but I couldn't spot any good ones, so I'll defer to you. 😅

File: plugins/bundle/plugin.go (hunk 2 above)
Comment (ashutosh-narkar) [reply plugins/bundle/plugin.go]: Done

File: download/download.go (hunk 1 above)
Comment (ashutosh-narkar) [reply download/download.go]: I was trying to push the limits here 😉 . Updated with a struct.

File: plugins/bundle/plugin_test.go (hunk 3 above)
Comment (srenatus) [reply plugins/bundle/plugin_test.go:2217]: Ignore me -- I thought this only did
```go
bundle.NewWriter(&amp;buf).UseModulePath(true).Write(b)
```
in a helper, but it's using `getTestBundle`, and wasn't meant to replace the similar calls above. We're good here. :D

**Review by srenatus — APPROVED** [review]
Thanks for addressing my comments!

**Pull request merged**